GAN/main.swift

```diff
@@ -0,0 +1,213 @@
+    return realLoss + fakeLoss
+}
+
+/// Returns a sample vector.
+func sampleVector(size: Int) -> Tensor<Float> {
+    Tensor<Float>(randomNormal: [size, latentSize])
+}
+
+// MNIST data logic
+func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
+    let start = index * batchSize
+    return x[start..<start+batchSize]
+}
+
+let images = readMNIST(imagesFile: "train-images-idx3-ubyte")
+
+var generator = Generator()
+var discriminator = Discriminator()
+
+let optG = Adam(for: generator, learningRate: 2e-4, beta1: 0.5)
+let optD = Adam(for: discriminator, learningRate: 2e-4, beta1: 0.5)
+
+// Noise vectors and plot function for testing
+let testImageGridSize = 4
+let testVector = sampleVector(size: testImageGridSize*testImageGridSize)
+func plotTestImage(_ testImage: Tensor<Float>, name: String) {
+    var imageGrid = testImage.reshaped(to: [testImageGridSize, testImageGridSize,
+                                            imageHeight, imageWidth])
+    
+    // Add padding.
+    imageGrid = imageGrid.padded(forSizes: [(0, 0), (0, 0), (1, 1), (1, 1)], with: 1)
+    
+    // Transpose to create single image.
+    imageGrid = imageGrid.transposed(withPermutations: [0, 2, 1, 3])
+    imageGrid = imageGrid.reshaped(to: [(imageHeight+2)*testImageGridSize,
+                                        (imageWidth+2)*testImageGridSize])
+    
+    // Convert [-1, 1] range to [0, 1] range.
+    imageGrid = (imageGrid + 1) / 2
+    
+    plot(image: imageGrid, name: name)
+}
+
+print("Start training...")
+
+// Start training loop.
+for epoch in 1...epochCount {
+    // Start training phase.
+    Context.local.learningPhase = .training
+    for i in 0 ..< Int(images.shape[0]) / batchSize {
+        // Perform alternative update.
+        // Update generator.
+        let vec1 = sampleVector(size: batchSize)
+        
+        let 𝛁generator = generator.gradient { generator -> Tensor<Float> in
+            let fakeImages = generator(vec1)
+            let fakeLogits = discriminator(fakeImages)
+            let loss = generatorLoss(fakeLogits: fakeLogits)
+            return loss
+        }
+        optG.update(&generator.allDifferentiableVariables, along: 𝛁generator)
+        
+        // Update discriminator.
+        let realImages = minibatch(in: images, at: i)
+        let vec2 = sampleVector(size: batchSize)
+        let fakeImages = generator(vec2)
+        
+        let 𝛁discriminator = discriminator.gradient { discriminator -> Tensor<Float> in
+            let realLogits = discriminator(realImages)
+            let fakeLogits = discriminator(fakeImages)
+            let loss = discriminatorLoss(realLogits: realLogits, fakeLogits: fakeLogits)
+            return loss
+        }
+        optD.update(&discriminator.allDifferentiableVariables, along: 𝛁discriminator)
+    }
+    
+    // Start inference phase.
+    Context.local.learningPhase = .inference
+    let testImage: Tensor<Float> = generator(testVector)
```

Comment:
```suggestion
    let testImage = generator(testVector)
```
Remove redundant type signature